### DISCUSSION TITLE
Multi-GPU error in Cusvaer

### DISCUSSION THREAD

**Discussion by suzuki-nd94hq** [NVIDIA/cuQuantum#87] [Q&A] [answered]
Hi,

I'm trying to execute on multi-GPU using cuQuantum appliance23.03 and cusvaer.
However, the following error occurs when performing 2 qubits state vector calculation for a GHZ state using 2 GPUs.
The same problem occurs with 3 qubits and 4 GPUs.
If you know any solutions, please let us know.

`$ mpirun -n 4 python ghz_statevec.py`
![image](https://github.com/NVIDIA/cuQuantum/assets/64938353/34f232a1-84d6-46c3-9b61-6df27445f11a)

The circuit being run is in the following GHZ state.
![image](https://github.com/NVIDIA/cuQuantum/assets/64938353/2dd2c577-a8e5-4f61-b45a-a9abad2e27fb)

Thanks in advance.

**Comment by mtjrider** [comment, marked as answer]
Hi @suzuki-nd94hq 

Thanks for posting your question. Yes, this behavior is expected.
The problem results from having only a single effective qubit state on each GPU while attempting to run a nominal two-qubit gate.

Given $2^{m}$ GPUs and a quantum circuit with $n$ qubits, (roughly speaking), we assume the following:

$$
\begin{aligned}
(1)\ \ & \log_{2}\left( 2^{n - m} \right) = k \\
(2)\ \ & k > g
\end{aligned}
$$

$(1)$ implies the size of any given GPU's _local_ quantum state is $k$ qubits. Further, we impose an additional requirement: the GPU's _local_ quantum state must accommodate the largest operation, with size represented by $g$ qubits. $g$ must also accommodate any gate-fusion being performed by the transpiler (e.g.) if you set `fusion_max_qubit=3`, then the largest operation will be $g=3$.

In the following cases:

1. a 2-qubit state on 2x GPUs running the GHZ circuit, $k=1; g=2$
2. a 3-qubit state on 4x GPUs running the GHZ circuit, $k=1; g=2$

there is no solution. You can either increase the size of the GPU's local state to accommodate the gate operations, or you can change the gate operations.

To elaborate on our reasoning: we didn't find any substantial benefit for running systems of this size in a distributed fashion on GPUs; however, if you have an application that would benefit from this pattern, please let us know.

For the future, we'll take the action of providing more descriptive warnings/errors for situations like this.
